### ISSUE TITLE
TimeZoneInfo.ToSerializedString/FromSerializedString do not round trip on Unix

### ISSUE THREAD

**Issue by justinvp** [dotnet/corefx#14797]
The following test passes on Windows but fails on any Unix:

```c#
public static IEnumerable<object[]> SystemTimeZonesTestData()
{
    foreach (TimeZoneInfo tz in TimeZoneInfo.GetSystemTimeZones())
    {
        yield return new object[] { tz };
    }
}

[Theory]
[MemberData(nameof(SystemTimeZonesTestData))]
public static void ToSerializedString_FromSerializedString_RoundTrips(TimeZoneInfo timeZone)
{
    string serialized = timeZone.ToSerializedString();
    TimeZoneInfo deserializedTimeZone = TimeZoneInfo.FromSerializedString(serialized);
    Assert.Equal(timeZone, deserializedTimeZone);
    Assert.Equal(serialized, deserializedTimeZone.ToSerializedString());
}
```

I believe it's due to how `GetAdjustmentRules` is implemented on Unix: https://github.com/dotnet/coreclr/blob/0a11492d52faa85c551761f8390be5de9d74e5ec/src/mscorlib/src/System/TimeZoneInfo.Unix.cs#L131-L153

vs. on Windows: https://github.com/dotnet/coreclr/blob/0a11492d52faa85c551761f8390be5de9d74e5ec/src/mscorlib/src/System/TimeZoneInfo.Win32.cs#L87

See #14795

**Comment by tarekgh** [comment]
I moved this to the future release because fixing this will need some major changes how we read and handle the adjustment rules in Linux which will be risky for v2. We should have a better story for time zones in general in the future.

**Comment by SUPERSC0TT** [comment]
Bumping to see if there are any plans to get this fixed. This bug means that users who create custom time zones via [microsoft's documentation](https://docs.microsoft.com/en-us/dotnet/standard/datetime/create-time-zones-with-adjustment-rules) will not be able to use GetAdjustmentRules() to see the correct definitions, namely the Transition Rules Start/End are not returned correctly.

**Comment by tarekgh** [comment]
@SUPERSC0TT there is no plan to address this very soon but we may get into that in the near future. the reason is we need to look at the whole TZ support and address the issues in more better designed way. 

Could you please share some more information about your TZ usage? like do you use custom TZ? why? what is your scenarios you are using with? 

We appreciate your feedback.

**Comment by yahorsi** [comment]
Guys, timezones are basic, we've got into this as well, and issues such this are a show-stopper if we need to move 2 the Linux ((

**Comment by tarekgh** [comment]
@yahorsi  could you please tell more about your scenario? why you want serialize TZI at all? are you creating your own custom TZ?

**Comment by yahorsi** [comment]
We're actually hit by aa bit another issue, the one that is described here:

https://github.com/dotnet/corefx/issues/17120

And caused by the following:
```
            // The rules we use in Unix care mostly about the start and end dates but don't fill the transition start and end info.
            // as the rules now is public, we should fill it properly so the caller doesn't have to know how we use it internally
            // and can use it as it is used in Windows
```

Our scenario is simple, we have a 'constant' custom Timezone that is parsed using TimeZoneInfo.FromSerializedString. In some parts of the code we just need to read transition start and end info, and it works on Linux and fails on Linux.

As I see in the code it should be very easy to fix. The question is why it was made in the first place.
As behaviour is really confusing. Seems like internally transition start and end info is correctly used  and only "fails" when you call GetAdjustmentRules

**Comment by tarekgh** [comment]
> As I see in the code it should be very easy to fix. The question is why it was made in the first place.

Internally the data stored inside adjustment rules in Linux is different than what we have in Windows. We have another issues (like the one you pointed at) tracking fixing exposing the rules correctly on Linux too. That is why you are seeing this issue. 
TZI originally designed around how Windows stored TZ data which is different than what Linux/IANA does which made us to workaround that by internally storing different data. There is some rules that cannot be easily expressed by what we expose in Adjustment rules and that is why there is some challenge here.

Will try to look at that in the next few weeks. I marked this issue for 5.0 release.

**Comment by yahorsi** [comment]
What people usually expect is that code just works the same way on the different operating systems. And we're hurt even if we didn't deploy on Linux as many today's build servers are using Linux and so, tests are just failing (

Is there any option to have the fix for the 2.2.* branch?

**Comment by tarekgh** [comment]
@danmosemsft @ericstj to advise regarding servicing this in previous releases. 

We don't have the fix yet and depend on the fix we can evaluate the risk. meanwhile, we may try to find you a workaround which may help parsing the Linux serialized TZ.

**Comment by yahorsi** [comment]
Sorry, do you mean it should be serialized different way for linux and windows?
Now it parses just fine, at least it does not throw, and seems time conversion from/to UTC/local works fine as well, what fails (wrong transition start and end info) is just GetAdjustmentRules call

**Comment by tarekgh** [comment]
> Sorry, do you mean it should be serialized in different way for linux and windows?

No. what I meant, if there is a way to detect the string is serialized on Linux and interpret the serialized data differently to get the original intended information. I don't know yet if this is possible. I just wanted to say, if there is any reasonable workaround can be used here.

**Comment by ericstj** [comment]
We do not accept servicing requests for past releases through Github, please see https://dotnet.microsoft.com/platform/support/policy/dotnet-core, https://github.com/dotnet/core/blob/master/microsoft-support.md for details around backporting to past releases.   The short answer is: if you need this for .NETCore 2.1, please open a [support request](https://support.microsoft.com/supportforbusiness/productselection/?sapid=4fd4947b-15ea-ce01-080f-97f2ca3c76e8), .NETCore 2.2 is no longer recieving hotfixes (only security fixes).

**Comment by yahorsi** [comment]
Thank you, we will have a look at possible workarounds and if needed I will create a proper fix request. BTW there is still time before 3.1 release, I'm afraid 5.0 is too far in the future

**Comment by ericstj** [comment]
There is still time for 3.1 but that window is closing quickly, we're already in "Ask mode" where we are scrutinizing risk of all fixes going into 3.1.  If we can come up with a low risk fix in the next week it has a chance at getting in.
